Feat apigw kinesis logging

RossMurr4y opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
* Enable APIGateway component to log access logs to the OpsData bucket with a Kinesis Firehose
* Implemented Kinesis Data Stream resources (unused, as weren't necessary in the end)
* Refactored WAF to use the same logging functionality as the API Gateway
   * This involved also updating WAF for `cdn` and `lb` components

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Would like to be able to process a large amount of invocation logs with a Kinesis Firehose. This is a part of a greater push to revisit the way logging works in general.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Locally tested and deployed successfully.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## Followup Actions
<!---
    Are the changes mandatory (breaking) or optional?
    What changes must a consumer of this repository make in order to utilise it?
    Are there other issues or steps that need to happen once this PR is merged?

    Add a checklist of items or leave the default of "None"
-->
- [ ] Requires https://github.com/hamlet-io/engine/pull/1417

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] None of the above.

ml019 commented on aws/components/lambda/state.ftl:
```diff
@@ -128,7 +128,8 @@
                     "authorise" : lambdaInvokePermission(id),
-                    "authorize" : lambdaInvokePermission(id)
+                    "authorize" : lambdaInvokePermission(id),
+                    "kinesis" : lambdaKinesisPermission(id)
```
Can we come up with a role which is less tied to the resource type e.g. "stream"?

roleyfoley replied:
Also need to make the distinction between the different kiniesis streams if we are adding support for datastreams. Whats the purpose of this role anyway?

ml019 reviewed (changes requested):
With merging of support for service to resource mapping, there will need to be a bit of extra config added for AWS_KINESIS_DATA_STREAM_RESOURCE_TYPE

roleyfoley commented on aws/components/s3/setup.ftl:
```diff
@@ -203,21 +202,11 @@
                     [/#if]
                     [#break]
 
+                [#case "baselinedata"]
```
Better to use the component type Variable instead of fixed value. Shows the meaning of what this is a bit more

roleyfoley commented on aws/services/kinesis/resource.ftl:
```diff
@@ -45,6 +84,149 @@
+
+        [/#switch]
+
+        [#-- [@createKinesisDataStream 
+            id=dataStream.Id
+            name=dataStream.Name
+            keyId=dataStreamEncryptionKey
+            dependencies=dependencies
+        /] --]
```
I don't think we should include kinesis streams as part of this process. This macro is to create a firehose not a data stream

RossMurr4y commented:
@ml019 @roleyfoley I've addressed your comments in 37d803b, 6c9ce96 & 3e4b92d.

I found the AWS documentation on Data and Delivery streams quite confusing and contradictory. 

- In the Kinesis Firehose documentation, the Delivery Streams discuss receiving data either by Direct PUT requests, or by Data Stream. There are also some places that used "Firehose Data Stream" and "Delivery Stream" interchangeably. This led me to believe that the Kinesis Data Stream was necessary, so I implemented it.
- In the API Gateway documentation, there are tutorials that have the APIGW communicate directly with the Firehose Delivery Stream. I couldn't get my tests to work however, but I could get the "test data" from the Firehose to the intended bucket.
- Eventually I worked out that the problem was that using the "test" calls within the API Gateway do not produce logs, so it was my testing and not implementation that was at fault.
- When I reversed some of what I'd implemented (dropping their git commits) it looks like some of that was caught up in other commits so was left behind (like the commented-out datastream.

So, I've left the macros for the creation of a Data Stream implemented should we want it in the future, but I've removed the remainder of my older implementations (including the lambda role).

RossMurr4y commented:
This has been updated to include refactorings for WAF logging, so that it uses the same implementation of implementing the Kinesis Firehose across components.

roleyfoley commented on aws/components/apigateway/id.ftl:
```diff
@@ -1,7 +1,18 @@
+            "Children" : [
+                {
+                    "Names" : "DestinationLink",
+                    "Description" : "Destination for the Execution logs. If not provided, Execution logs will be stored in the OpsData DataBucket.",
```
```suggestion
                    "Description" : "Destination for the Access logs. If not provided, Execution logs will be stored in the OpsData DataBucket.",
```

RossMurr4y replied:
Thanks, have amended db4ed004

roleyfoley commented on aws/components/cdn/setup.ftl:
```diff
@@ -478,11 +478,23 @@
-            occurrence=occurrence
+    
+        [#local wafFirehoseStreamId = 
+            formatResourceId(AWS_KINESIS_FIREHOSE_STREAM_RESOURCE_TYPE, wafAclId)]
+
+        [@setupFirehoseStream
+            id=wafFirehoseStreamId
+            lgPath=formatAbsolutePath(core.FullAbsolutePath, "waf")
+            destinationLink=baselineLinks["OpsData"]
+            cmkKeyId=kmsKeyId
+            bucketPrefix=formatRelativePath(occurrence.Core.FullRelativePath, "waf")
+            errorPrefix=formatRelativePath(occurrence.Core.FullRelativePath, "waf", "error")
+            streamNamePrefix="aws-waf-logs-"
+        /]
+
+        [@createWAFLoggingDeliveryStream
             wafaclId=wafAclId
-            loggingProfile=wafLoggingProfile
+            deliveryStreamId=wafFirehoseStreamId
             regional=false
         /]
 
```
Should this check for `[#if solution.WAF.Logging.Enabled`

RossMurr4y replied:
Yep thanks, fixed up in 1cf5d5be

roleyfoley commented on aws/components/lb/setup.ftl:
```diff
@@ -722,6 +722,25 @@
         [#case "application"]
             [#if wafAclResources?has_content ]
 
+                [#local wafFirehoseStreamId = 
+                    formatResourceId(AWS_KINESIS_FIREHOSE_STREAM_RESOURCE_TYPE, wafAclResources.acl.Id)]
+
+                [@setupFirehoseStream
+                    id=wafFirehoseStreamId
+                    lgPath=formatAbsolutePath(core.FullAbsolutePath, "waf")
+                    destinationLink=baselineLinks["OpsData"]
+                    cmkKeyId=kmsKeyId
+                    bucketPrefix=formatRelativePath(occurrence.Core.FullRelativePath, "waf")
+                    errorPrefix=formatRelativePath(occurrence.Core.FullRelativePath, "waf", "error")
+                    streamNamePrefix="aws-waf-logs-"
+                /]
+
+                [@createWAFLoggingDeliveryStream
+                    wafaclId=wafAclResources.acl.Id
+                    deliveryStreamId=wafFirehoseStreamId
+                    regional=false
+                /]
```
Should this check for `[#if solution.WAF.Logging.Enabled`

RossMurr4y replied:
Corrected in 03237109, cheers

ml019 commented on aws/components/s3/setup.ftl:
```diff
@@ -202,6 +202,7 @@
                     [/#if]
                     [#break]
 
+                [#case BASELINE_DATA_COMPONENT_TYPE ]
```
```suggestion
```

ml019 commented on aws/components/apigateway/setup.ftl:
```diff
@@ -252,15 +260,37 @@
-    [#local accessLgId   = resources["accesslg"].Id]
-    [#local accessLgName = resources["accesslg"].Name]
+    [#if solution.AccessLogging.Enabled]
+        [#-- Manage Access Logs with Kinesis Firehose --]
```
Need an extra attribute to control whether firehose or cloud watch log groups is used. Also need logic to permit existing installations to preserve the existing cloudwatch log groups.

So I think something like `"aws:Firehose: True"` and `"aws:KeepLogGroup" : true`. stageLogTarget is controlled via the former. The creation of the log group is performed if either Firehose=false or KeepLogGroup=true.

Then we can incrementally remove the log groups once the data has been processed via firehose as a separate activity.

ml019 commented on aws/components/apigateway/id.ftl:
```diff
@@ -1,7 +1,18 @@
+    attributes=[
+        {
+            "Names" : "AccessLogging",
+            "Children" : [
```
See below re adding a couple of extra children to switch between firehose and log groups.

RossMurr4y replied:
@ml019 I have incorporated your changes. Could you take another look?

roleyfoley reviewed (approved):
All requested fixes added
